main.cpp

```diff
@@ -1838,44 +1838,61 @@ bool load_command::execute(device_map &devices) {
             }
         }
     }
-    for (auto mem_range : ranges) {
-        enum memory_type type = get_memory_type(mem_range.from);
-        // new scope for progress bar
-        {
-            progress_bar bar("Loading into " + memory_names[type] + ": ");
-            uint32_t batch_size = FLASH_SECTOR_ERASE_SIZE;
-            bool ok = true;
+
+    size_t total_size = 0;
+    for (auto mem_range : ranges) { total_size += mem_range.to - mem_range.from; }
+
+    {
+        progress_bar bar("Loading:   ");
+        size_t loaded_size = 0;
+        for (auto mem_range : ranges) {
+            enum memory_type type = get_memory_type(mem_range.from);
             vector<uint8_t> file_buf;
             vector<uint8_t> device_buf;
-            for (uint32_t base = mem_range.from; base < mem_range.to && ok; ) {
+            for (uint32_t base = mem_range.from; base < mem_range.to; ) {
+                uint32_t batch_size = FLASH_SECTOR_ERASE_SIZE;
+                uint8_t block_erase_cmd = 0x20;
+                if (mem_range.to - base >= 0x10000) {
+                    batch_size = 0x10000;
+                    block_erase_cmd = 0xD8;
+                }
+                else if (mem_range.to - base >= 0x8000) {
+                    batch_size = 0x8000;
+                    block_erase_cmd = 0x52;
+                }
```

Comment:
Probably makes sense to add `#define` values for these different erase cmds?
(but like you, I dunno how widely supported these commands are amongst other flash chips, and that'll be the kicker here!)